File: src/main/java/rag/data/ChunkLoader.java
package main.java.rag.data;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import main.java.rag.model.Chunk;


public class ChunkLoader {
    
    
    public ChunkStore loadChunks(Path chunksJsonPath) {
        try {
            List<String> lines = Files.readAllLines(chunksJsonPath);
            String jsonContent = String.join("\n", lines);
            
            ChunkStore chunkStore = new ChunkStore();
            parseJsonAndLoadChunks(jsonContent, chunkStore);
            
            return chunkStore;
        } catch (IOException e) {
            throw new RuntimeException("Failed to load chunks from: " + chunksJsonPath, e);
        }
    }
    

    private void parseJsonAndLoadChunks(String jsonContent, ChunkStore chunkStore) {
        // Remove whitespace and newlines for easier parsing
        String cleaned = jsonContent.replaceAll("\\s+", " ").trim();
        
        // Find documents array
        int documentsStart = cleaned.indexOf("\"documents\"");
        if (documentsStart == -1) {
            throw new RuntimeException("Invalid JSON: 'documents' key not found");
        }
        
        int arrayStart = cleaned.indexOf("[", documentsStart);
        if (arrayStart == -1) {
            throw new RuntimeException("Invalid JSON: documents array not found");
        }
        
        // Parse each document
        int pos = arrayStart + 1;
        while (pos < cleaned.length()) {
            int docStart = cleaned.indexOf("{", pos);
            if (docStart == -1) break;
            
            int docEnd = findMatchingBrace(cleaned, docStart);
            if (docEnd == -1) break;
            
            String docJson = cleaned.substring(docStart, docEnd + 1);
            parseDocument(docJson, chunkStore);
            
            pos = docEnd + 1;
        }
    }
    
   
    private void parseDocument(String docJson, ChunkStore chunkStore) {
        // Extract docId
        String docId = extractStringValue(docJson, "docId");
        if (docId == null) {
            throw new RuntimeException("Invalid document: docId not found");
        }
        
        // Extract title
        String title = extractStringValue(docJson, "title");
        if (title != null) {
            chunkStore.setDocumentTitle(docId, title);
        }
        
        // Extract sections array
        int sectionsStart = docJson.indexOf("\"sections\"");
        if (sectionsStart == -1) return;
        
        int arrayStart = docJson.indexOf("[", sectionsStart);
        if (arrayStart == -1) return;
        
        // Parse each section
        int pos = arrayStart + 1;
        while (pos < docJson.length()) {
            int sectionStart = docJson.indexOf("{", pos);
            if (sectionStart == -1) break;
            
            int sectionEnd = findMatchingBrace(docJson, sectionStart);
            if (sectionEnd == -1) break;
            
            String sectionJson = docJson.substring(sectionStart, sectionEnd + 1);
            parseSection(sectionJson, docId, chunkStore);
            
            pos = sectionEnd + 1;
        }
    }
    
    
    private void parseSection(String sectionJson, String docId, ChunkStore chunkStore) {
        // Extract sectionId
        String sectionId = extractStringValue(sectionJson, "sectionId");
        if (sectionId == null) {
            sectionId = ""; // Default if not found
        }
        
        // Extract chunks array
        int chunksStart = sectionJson.indexOf("\"chunks\"");
        if (chunksStart == -1) return;
        
        int arrayStart = sectionJson.indexOf("[", chunksStart);
        if (arrayStart == -1) return;
        
        // Parse each chunk
        int pos = arrayStart + 1;
        while (pos < sectionJson.length()) {
            int chunkStart = sectionJson.indexOf("{", pos);
            if (chunkStart == -1) break;
            
            int chunkEnd = findMatchingBrace(sectionJson, chunkStart);
            if (chunkEnd == -1) break;
            
            String chunkJson = sectionJson.substring(chunkStart, chunkEnd + 1);
            parseChunk(chunkJson, docId, sectionId, chunkStore);
            
            pos = chunkEnd + 1;
        }
    }
    
    
    private void parseChunk(String chunkJson, String docId, String sectionId, ChunkStore chunkStore) {
        String chunkId = extractStringValue(chunkJson, "chunkId");
        String content = extractStringValue(chunkJson, "content");
        int startOffset = extractIntValue(chunkJson, "startOffset");
        int endOffset = extractIntValue(chunkJson, "endOffset");
        
        if (chunkId == null || content == null) {
            throw new RuntimeException("Invalid chunk: missing required fields");
        }
        
        Chunk chunk = new Chunk(docId, chunkId, content, sectionId, startOffset, endOffset);
        chunkStore.addChunk(chunk);
    }
    
    
    private String extractStringValue(String json, String key) {
        String searchKey = "\"" + key + "\"";
        int keyPos = json.indexOf(searchKey);
        if (keyPos == -1) return null;
        
        int colonPos = json.indexOf(":", keyPos);
        if (colonPos == -1) return null;
        
        // Find the value (string between quotes)
        int quoteStart = json.indexOf("\"", colonPos);
        if (quoteStart == -1) return null;
        
        int quoteEnd = json.indexOf("\"", quoteStart + 1);
        if (quoteEnd == -1) return null;
        
        return json.substring(quoteStart + 1, quoteEnd);
    }
    
    /**
     * Extract integer value from JSON by key.
     */
    private int extractIntValue(String json, String key) {
        String searchKey = "\"" + key + "\"";
        int keyPos = json.indexOf(searchKey);
        if (keyPos == -1) return 0;
        
        int colonPos = json.indexOf(":", keyPos);
        if (colonPos == -1) return 0;
        
        // Find the number
        int numStart = colonPos + 1;
        while (numStart < json.length() && Character.isWhitespace(json.charAt(numStart))) {
            numStart++;
        }
        
        int numEnd = numStart;
        while (numEnd < json.length() && (Character.isDigit(json.charAt(numEnd)) || json.charAt(numEnd) == '-')) {
            numEnd++;
        }
        
        try {
            return Integer.parseInt(json.substring(numStart, numEnd).trim());
        } catch (NumberFormatException e) {
            return 0;
        }
    }
    
    /**
     * Find matching closing brace for an opening brace.
     */
    private int findMatchingBrace(String str, int openBracePos) {
        int depth = 1;
        int pos = openBracePos + 1;
        
        while (pos < str.length() && depth > 0) {
            char c = str.charAt(pos);
            if (c == '{') {
                depth++;
            } else if (c == '}') {
                depth--;
            } else if (c == '"') {
                // Skip string content
                pos++;
                while (pos < str.length() && str.charAt(pos) != '"') {
                    if (str.charAt(pos) == '\\') {
                        pos++; // Skip escaped character
                    }
                    pos++;
                }
            }
            pos++;
        }
        
        return depth == 0 ? pos - 1 : -1;
    }
}

File: iteration1_py/src/data/chunk_loader.py
import json
from pathlib import Path
from typing import Optional

from src.model.chunk import Chunk
from src.data.chunk_store import ChunkStore


class ChunkLoader:
   
    def load_chunks(self, chunks_json_path: Path) -> ChunkStore:
       
        chunks_json_path = Path(chunks_json_path)
        
        if not chunks_json_path.exists():
            raise FileNotFoundError(f"Chunks file not found: {chunks_json_path}")
        
        try:
            with open(chunks_json_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            chunk_store = ChunkStore()
            self._parse_and_load_chunks(data, chunk_store)
            
            return chunk_store
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON in file {chunks_json_path}: {e}")
        except Exception as e:
            raise RuntimeError(f"Failed to load chunks from {chunks_json_path}: {e}")
    
    @staticmethod
    def _parse_and_load_chunks(data: dict, chunk_store: ChunkStore) -> None:
       
        if 'documents' not in data:
            raise ValueError("Invalid JSON structure: 'documents' key not found")
        
        documents = data['documents']
        
        for document in documents:
            ChunkLoader._parse_document(document, chunk_store)
    
    @staticmethod
    def _parse_document(doc_data: dict, chunk_store: ChunkStore) -> None:
        
        doc_id = doc_data.get('docId')
        if doc_id is None:
            raise ValueError("Invalid document: docId not found")
        
        title = doc_data.get('title')
        if title:
            chunk_store.set_document_title(doc_id, title)
        
        sections = doc_data.get('sections', [])
        for section in sections:
            ChunkLoader._parse_section(section, doc_id, chunk_store)
    
    @staticmethod
    def _parse_section(section_data: dict, doc_id: str, chunk_store: ChunkStore) -> None:
       
        section_id = section_data.get('sectionId', '')
        chunks = section_data.get('chunks', [])
        
        for chunk_data in chunks:
            ChunkLoader._parse_chunk(chunk_data, doc_id, section_id, chunk_store)
    
    @staticmethod
    def _parse_chunk(chunk_data: dict, doc_id: str, section_id: str, chunk_store: ChunkStore) -> None:
        
        chunk_id = chunk_data.get('chunkId')
        content = chunk_data.get('content')
        
        if not chunk_id or not content:
            raise ValueError("Invalid chunk: missing required fields (chunkId, content)")
        
        start_offset = chunk_data.get('startOffset', 0)
        end_offset = chunk_data.get('endOffset', 0)
        
        chunk = Chunk(
            doc_id=doc_id,
            chunk_id=chunk_id,
            text=content,
            section_id=section_id,
            start_offset=start_offset,
            end_offset=end_offset
        )
        
        chunk_store.add_chunk(chunk)
